### PR TITLE
Add support for x509 SubjectAltName identification

### PR DESCRIPTION
**Pull request by tonickkozlov**
Currently only CommonName of an X.509 certificate can be used to validate identity of a user.
This change extends support to include [SubjectAltName extension](https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.6) as well.

With this change, a user can be defined as as 
```
<user>
    <ssl_certificates>
        <subject_alt_name>URI:spiffe://foo.com/bar</subject_alt_name>
    </ssl_certificates>
</user>
```

Then a user would need to present the following info in its TLS certificate:
```
Certificate:
  Data: ...
    X509v3 extensions:
        X509v3 Subject Alternative Name:
            URI:spiffe://foo.com/bar
 ```

Few caveats:
- Only DNS and URI types of subjectAltName extension are supported
- Wildcard validation is not supported, it's not required by the standard

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add support for user identification based on x509 SubjectAltName extension.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit fd54874d4e4d7ac93a526fa50e91addbb5680a7d with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65626/fd54874d4e4d7ac93a526fa50e91addbb5680a7d/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Cloud fork sync (only for ClickHouse Inc. employees)</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td><a href="https://github.com/ClickHouse/ClickHouse/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1">❌ failure</a></td></tr>
<tbody>
</table>

File: src/Access/AuthenticationData.cpp [hunk 1]
```diff
@@ -15,6 +15,7 @@
 #include <boost/algorithm/hex.hpp>
 #include <boost/algorithm/string/case_conv.hpp>
 
+#include "Access/Common/SSLCertificateSubjects.h"
```
Comment (thevar1able) [inline src/Access/AuthenticationData.cpp:18]:
```suggestion
#include <Access/Common/SSLCertificateSubjects.h>
```

File: docs/en/operations/external-authenticators/ssl-x509.md [hunk 2]
```diff
@@ -6,23 +6,30 @@ import SelfManaged from '@site/docs/en/_snippets/_self_managed_only_no_roadmap.m
...
+[SSL 'strict' option](../server-configuration-parameters/settings.md#server_configuration_parameters-openssl) enables mandatory certificate validation for the incoming connections. In this case, only connections with trusted certificates can be established. Connections with untrusted certificates will be rejected. Thus, certificate validation allows to uniquely authenticate an incoming connection. `Common Name` or `subjectAltName extension` field of the certificate are used to identify connected user. This allows to associate multiple certificates with the same user. Additionally, reissuing and revoking of the certificates does not affect the ClickHouse configuration.
```
Comment (thevar1able) [inline docs/en/operations/external-authenticators/ssl-x509.md:9]:
```suggestion
[SSL 'strict' option](../server-configuration-parameters/settings.md#server_configuration_parameters-openssl) enables mandatory certificate validation for the incoming connections. In this case, only connections with trusted certificates can be established. Connections with untrusted certificates will be rejected. Thus, certificate validation allows to uniquely authenticate an incoming connection. `Common Name` or `subjectAltName extension` field of the certificate is used to identify the connected user. This allows to associate multiple certificates with the same user. Additionally, reissuing and revoking of the certificates does not affect the ClickHouse configuration.
```

File: src/Server/HTTPHandler.cpp [hunk 3]
```diff
@@ -24,6 +24,7 @@
 #include <Server/HTTPHandlerFactory.h>
 #include <Server/HTTPHandlerRequestFilter.h>
 #include <Server/IServer.h>
+#include "Common/OpenSSLHelpers.h"
```
Comment (thevar1able) [inline src/Server/HTTPHandler.cpp:27]:
```suggestion
#include <Common/OpenSSLHelpers.h>
```

**Comment by tonickkozlov** [comment]
@thevar1able thanks for taking a look at my PR. Does the overall approach look OK to you?

> 00626_replace_partition_from_table_zookeeper
This test failed with a timeout but it doesn't seem related to the change I'm making. Is there a chance it's flaky?

**Comment by thevar1able** [comment]
@tonickkozlov yep overall looks OK, I'll make another pass today. I will check the test as well.

**Review by thevar1able — APPROVED** [review]

**Pull request merged**